### PR TITLE
[python] disabled early stopping in dart mode

### PR DESCRIPTION
**Pull request by StrikerRUS**
Refer to https://github.com/Microsoft/LightGBM/issues/1893#issuecomment-444803315.

Maybe error is better, because users tend to completely ignore warnings?

### CODE REVIEW

**Comment by guolinke** [comment]
@StrikerRUS I think we can disable early_stoping when using dart, even when early_stopping is set.

**Comment by StrikerRUS** [comment]
@guolinke At cpp side?

**Comment by guolinke** [comment]
@StrikerRUS the change of cpp side is in another PR.
The logic of early stopping is independent in cpp/python/R, so we need to disable it in python side as well.

**Comment by StrikerRUS** [comment]
@guolinke Ah, I see!

**Comment by StrikerRUS** [comment]
@guolinke I've updated PR. Unfortunately, I'm not familiar with `callback` module yet, therefore this solution might be ugly. Please check.

File: python-package/lightgbm/callback.py [hunk 1]
```diff
@@ -198,8 +199,17 @@ def _init(env):
                 cmp_op.append(lt)
 
     def _callback(env):
+        enabled = not any((boost_alias in env.params
+                           and env.params[boost_alias] == 'dart') for boost_alias in ('boosting',
+                                                                                      'boosting_type',
+                                                                                      'boost'))
```
Comment (guolinke) [inline python-package/lightgbm/callback.py]: maybe this can be moved into _init(env) ?

Comment (StrikerRUS) [reply python-package/lightgbm/callback.py]: @guolinke Sorry, I was in a hurry and mislooked [this aspect](https://stackoverflow.com/questions/5218895/python-nested-functions-variable-scoping), which made me think that it's impossible to set `enabled` variable in `_init` function. I've reworked this.

**Comment by Teeeto** [comment]
Sometimes early stopping is needed, disabling it may render dart mode quite useless. I'd suggest to save/override the best model (or at least a mask for dropped trees) in memory at each iteration and return the proper model when the early stopping is triggered.

**Pull request merged**